

# Isoform-specific perturb-seq reveals alternative promoter dynamics in gene regulation and drug response

## Abstract
CRISPR-dCas9 technologies are designed to modulate gene expression. However, our findings reveal that many hits are promoter and isoform-specific. Leveraging this specificity, we developed an isoform-specific single-cell Perturb-Seq screen, showing that alternative promoters frequently compensate for targeted knockdowns and drive divergent gene expression programs at the single-cell level. Importantly, these isoform-specific effects influence drug response in breast cancer, underscoring the relevance of alternative promoters in functional genomics and drug discovery.

![Visual Figure](./source/fig/VisualAbstract_vNAR_v2.png)


## Links
Please access the website: 

- https://isoform-specific-perturb-seq.readthedocs.io/en/latest/index.html

Please access the raw data: 

- https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-14567?key=244774e5-fc31-4e23-8102-4fbd49b2ed09

Please access the paper:

- https://academic.oup.com/nar/article/54/4/gkag118/8494758 


## Pipeline

***1_Promoter_identification***
A. Integration of RNA-seq, ChIP-seq and CAGE-seq for MCF-7 
B. Identify the genes with ideal APU events for guide 
C. Comparison of promoter calling against refTSS and EPD

***2_Guide_design***
2.  Choosing the ideal promoter-specific guide design 

***3_Analysis_Scripts***

**APU_ANALYSIS** 
Package that contains enclosed functions. 

1. Run through the 10x Cellranger pipeline and velocyto for single cell RNAseq quantification and using (2) guides quantification. all found in the cellranger files folder **bash**
2.  Guide Calling for dual guide. Use repogle method to take molecule.h5 generated by cellranger and py to run through repogle version of guide calling or use cellranger_guidecalling.ipynb for Direct Capture Perturb-Seq dual guide. Formed guide-specific lists of cells.
3. Pseudobulk analysis.
    A. Separation of guide-specific FASTQ files. **bash**
    B. Whippet pseudobulk is used for transcript-specific analysis, post-UMI deduplication.  **bash**
    C. Transcript quality control. **R**
    D. Whippet result visualisation.
4. Normalisation of adata object and E-distance of KD
5. Check gene and neighbouring gene expression
6. Create individual UMAPS per gene of interest 
    A. UMAPs 
    B. Rand Index score
7. Cell phase assignment model from FUCCI-matched single cell paper (GSE146773)
8. Differential Expression analysis.
    A. Find the shared P1 and P2 genes. 
    B. Check the shared P1 and P2 across protospacers with the same A/B and C/D.
9.  CNV Score & Numbat to quantify and Velocity quantification with loom file
10. ESR1-specific analysis from proliferation analysis to rt-qpcr
11.  Spectra analysis and visualisation for pathway enrichment
