

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Quantitative individual UMAPs per gene of interest &mdash; Isoform Specific Perturb-Seq  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=9edc463e" />
      <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
      <link rel="stylesheet" type="text/css" href="../../_static/nbsphinx-code-cells.css?v=2aa19091" />

  
      <script src="../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../../_static/doctools.js?v=fd6eb6e6"></script>
      <script src="../../_static/sphinx_highlight.js?v=6ffebe34"></script>
      <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
      <script src="../../_static/copybutton.js?v=f281be69"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@4/tex-mml-chtml.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Cell Assignment Model from FUCCI-Matched Single-Cell Paper" href="7_cellphase_model.html" />
    <link rel="prev" title="Target gene and neighbouring gene expression" href="5_genekd_neighbouring_gene_expression.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            Isoform Specific Perturb-Seq
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Analysis Pipeline:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../promoter_identification.html">Promoter Identification</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../guide_design.html">Guide Design</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../analysis_scripts.html">Analysis Scripts &amp; Pipelines</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../analysis_scripts.html#preprocessing-guide-calling">Preprocessing &amp; Guide Calling</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../analysis_scripts.html#splicing-rna-velocity">Splicing &amp; RNA Velocity</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../analysis_scripts.html#transcriptional-divergence-e-stats">Transcriptional Divergence (E-Stats)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../analysis_scripts.html#cell-phase-pathway-modeling">Cell Phase &amp; Pathway Modeling</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../analysis_scripts.html#clinical-differential-expression">Clinical &amp; Differential Expression</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="2A_cellranger_guidecalling.html">Guide Calling for Dual Guides</a></li>
<li class="toctree-l3"><a class="reference internal" href="2B_guide_calling_cellrangerfilter4_idealguides.html">Filtering for Cells with Ideal Combination of Guides</a></li>
<li class="toctree-l3"><a class="reference internal" href="4_estatistic_gene_guide.html">Guide Calling Using the Cellranger Module HDF5 File</a></li>
<li class="toctree-l3"><a class="reference internal" href="5_genekd_neighbouring_gene_expression.html">Target gene and neighbouring gene expression</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Quantitative individual UMAPs per gene of interest</a></li>
<li class="toctree-l3"><a class="reference internal" href="7_cellphase_model.html">Cell Assignment Model from FUCCI-Matched Single-Cell Paper</a></li>
<li class="toctree-l3"><a class="reference internal" href="8_perturbseq_differentialexpression.html">Differential Expression analysis between P1 &amp; P2</a></li>
<li class="toctree-l3"><a class="reference internal" href="9A_cnv_score.html">Investigating Copy Number Variation Across KD Populations</a></li>
<li class="toctree-l3"><a class="reference internal" href="9B_velocity_loom.html">Velocity and pyGSEA calculations of ESR1</a></li>
<li class="toctree-l3"><a class="reference internal" href="11_spectra.html">Spectra Analysis</a></li>
<li class="toctree-l3"><a class="reference internal" href="12_negativecontrol.html">Guide Calling Using the Cellranger Module HDF5 File</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Isoform Specific Perturb-Seq</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../analysis_scripts.html">Analysis Scripts &amp; Pipelines</a></li>
      <li class="breadcrumb-item active">Quantitative individual UMAPs per gene of interest</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/scripts/3_analysis_scripts/6_perturbseq_tsne_kldivergence_umap.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Quantitative-individual-UMAPs-per-gene-of-interest">
<h1>Quantitative individual UMAPs per gene of interest<a class="headerlink" href="#Quantitative-individual-UMAPs-per-gene-of-interest" title="Link to this heading"></a></h1>
<ol class="arabic simple">
<li><p>Run through the 10x Cellranger pipeline and velocyto for single cell RNAseq quatification and using (2) guides quantifiction. all found in the cellranger files folder <strong>bash</strong></p></li>
<li><p>Guide Calling for dual guide. Use repogle method to take molecule.h5 generated by cellranger and py to run through repogle version of guide calling or use cellranger_guidecalling.ipynb for Direct Capture Perturb-Seq dual guide. Formed guide-specific lists of cells.</p></li>
<li><p>Pseudobulk analysis. A. Seperation of guide-specific fastq files. <strong>bash</strong> B. Whippet pseudobulk for transcript specific analysis, post UMI deduplication. <strong>bash</strong> C. Transcript quality control. <strong>R</strong> D. Whippet result visualisation.</p></li>
<li><p>Normalisation of adata object and E-distance of KD</p></li>
<li><p>Check gene and neighboring gene expression</p></li>
<li><p><strong>Create individual umaps per gene of interest</strong> A. UMAPs B. Rand Index score</p></li>
<li><p>Cell phase assignment model from FUCCI-matched single cell paper (GSE146773_)</p></li>
<li><p>Differential Expression analysis. A. Find the shared P1 and P2 genes. B. Check the shared P1 and P2 across different protospacers with the same A/B and C/D.</p></li>
<li><p>CNV Score &amp; Numbat to quantify and Velocity quantification with loom file</p></li>
<li><p>ESR1-specific analysis from proliferation analysis to rt-qpcr</p></li>
<li><p>Spectra analysis and visualisation for pathway enrichment</p></li>
</ol>
<p>This script focuses on using unsupervised learning and information theory to prove that alternative promoters drive distinct transcriptional states.</p>
<p>Visualization of Transcriptional Landscapes (UMAP &amp; t-SNE) The primary goal of this notebook is to visualize how cells cluster based on which promoter (P1 or P2) was targeted.</p>
<p>What is happening: The code performs high-dimensional reduction using UMAP (Uniform Manifold Approximation and Projection) and t-SNE (t-distributed Stochastic Neighbor Embedding).</p>
<p>The Logic: If the two promoters were redundant, their respective cell populations would overlap completely. Instead, for genes like ESR1, BRIP1, and MYBB1A, the notebook generates plots showing cells segregating into distinct “clouds” or gradients based on the promoter knockdown</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">load_ext</span> autoreload
<span class="o">%</span><span class="k">matplotlib</span> inline
<span class="o">%</span><span class="k">autoreload</span> 2
<span class="c1">#general</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pl</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ad</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">hdf5plugin</span>

<span class="c1">#form a location</span>
<span class="n">loc</span><span class="o">=</span><span class="s2">&quot;alt-prom-crispr-fiveprime/&quot;</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">tqdm.notebook</span><span class="w"> </span><span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">loc</span><span class="o">+</span><span class="s1">&#39;scripts/&#39;</span><span class="p">)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">apu_analysis</span><span class="w"> </span><span class="kn">import</span> <span class="o">*</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scperturb</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">infercnvpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">cnv</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">apu_analysis.cell_import</span><span class="w"> </span><span class="kn">import</span> <span class="n">CellPopulation</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">IPython.display</span><span class="w"> </span><span class="kn">import</span> <span class="n">clear_output</span>
<span class="n">pd</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">display</span><span class="o">.</span><span class="n">float_format</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{:.4f}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>

<span class="c1">#for this python</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.special</span><span class="w"> </span><span class="kn">import</span> <span class="n">rel_entr</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sklearn.cluster</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">cluster</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">umap</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy</span><span class="w"> </span><span class="kn">import</span> <span class="n">stats</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.stats</span><span class="w"> </span><span class="kn">import</span> <span class="n">bootstrap</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">statsmodels.api</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sm</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">statsmodels.stats.multitest</span><span class="w"> </span><span class="kn">import</span> <span class="n">multipletests</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="kn">import</span> <span class="n">reshape</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="kn">import</span> <span class="n">array</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.decomposition</span><span class="w"> </span><span class="kn">import</span> <span class="n">PCA</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">hdbscan</span>

<span class="c1"># Taken from:</span>
<span class="c1"># Adamson, B.A., Norman, T.M., *et al.* &quot;A multiplexed CRISPR screening platform enables systematic dissection of the unfolded protein response&quot;, *Cell*, 2016.</span>
<span class="c1"># My experiment deals with two KDs- one of the MP, one of the AP using two guides. Positve controls include GINS1 ect. This is a combnatorial KD double for the same gene. No treatments were used</span>

<span class="c1"># colours using garvan</span>
<span class="n">color1</span> <span class="o">=</span><span class="s1">&#39;#4d00c7&#39;</span>
<span class="n">palecolor1</span><span class="o">=</span><span class="s2">&quot;#b366ff&quot;</span>
<span class="n">color2</span><span class="o">=</span> <span class="s1">&#39;#da3c07&#39;</span>
<span class="n">palecolor2</span><span class="o">=</span><span class="s2">&quot;#ff8954&quot;</span>
<span class="n">color3</span><span class="o">=</span><span class="s1">&#39;#05d3d3&#39;</span>
<span class="n">color4</span><span class="o">=</span><span class="s1">&#39;#c6c7c5&#39;</span>
<span class="n">color4</span><span class="o">=</span><span class="s2">&quot;#434541&quot;</span>
<span class="n">color5</span><span class="o">=</span><span class="s2">&quot;#eb31e1&quot;</span>
<span class="n">color6</span><span class="o">=</span><span class="s2">&quot;#3175eb&quot;</span>
<span class="n">color7</span><span class="o">=</span><span class="s2">&quot;#a7eb31&quot;</span>
<span class="n">color8</span><span class="o">=</span><span class="s2">&quot;#b366ff&quot;</span>
<span class="n">color9</span><span class="o">=</span><span class="s2">&quot;#ff8954&quot;</span>
<span class="n">color10</span><span class="o">=</span><span class="s2">&quot;#35c9d4&quot;</span>

<span class="c1">#use viridis</span>
<span class="n">color1</span><span class="o">=</span><span class="s2">&quot;#fde725&quot;</span>
<span class="n">color2</span><span class="o">=</span><span class="s2">&quot;#7ad151&quot;</span>
<span class="n">color3</span><span class="o">=</span><span class="s2">&quot;#22a884&quot;</span>
<span class="n">color4</span><span class="o">=</span><span class="s2">&quot;#2a788e&quot;</span>
<span class="n">color5</span><span class="o">=</span><span class="s2">&quot;#2a788e&quot;</span>
<span class="n">color6</span><span class="o">=</span><span class="s1">&#39;#440154&#39;</span>
<span class="c1"># %%capture</span>


<span class="c1"># Create the color palette</span>
<span class="n">palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">([</span><span class="n">palecolor1</span><span class="p">,</span><span class="n">palecolor2</span><span class="p">])</span>
<span class="n">palette2</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">([</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span> <span class="n">color3</span><span class="p">,</span> <span class="n">color4</span><span class="p">,</span><span class="n">color5</span><span class="p">,</span><span class="n">color6</span> <span class="p">,</span><span class="n">color7</span><span class="p">])</span>
<span class="n">umap_palette</span><span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">([</span><span class="s2">&quot;#F05325&quot;</span><span class="p">,</span><span class="s2">&quot;#04AFBB&quot;</span><span class="p">])</span>
<span class="c1"># Create the color palette</span>
<span class="n">palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">([</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span><span class="n">color3</span><span class="p">])</span>
<span class="n">new_palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">([</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span> <span class="n">color3</span><span class="p">,</span> <span class="n">color4</span><span class="p">])</span>

<span class="c1">#change palette to 3d82c4 36a047</span>
<span class="n">palette</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">([</span><span class="s2">&quot;#f3766e&quot;</span><span class="p">,</span> <span class="s2">&quot;#7094cd&quot;</span><span class="p">])</span>
<span class="c1">#change #36a047 to G1 #7671b3 toG2M and #d76127 to S</span>
<span class="n">cellcycle_palette</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">([</span><span class="s2">&quot;#36a047&quot;</span><span class="p">,</span> <span class="s2">&quot;#7671b3&quot;</span><span class="p">,</span><span class="s2">&quot;#d76127&quot;</span><span class="p">])</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Scanpy&quot;</span><span class="p">,</span> <span class="n">sc</span><span class="o">.</span><span class="n">__version__</span><span class="p">)</span>
<span class="o">%</span><span class="k">matplotlib</span> inline
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/helenking/anaconda3/envs/apu/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Scanpy 1.10.3
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%capture</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">loc</span><span class="o">+</span><span class="s2">&quot;files/adata_normalised.h5ad&quot;</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;log1p&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">plt</span><span class="o">.</span><span class="n">rc_context</span><span class="p">():</span>  <span class="c1"># Use this to set figure params like size and dpi</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">loc</span><span class="o">+</span><span class="s2">&quot;figures/hvg.pdf&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
<span class="k">with</span> <span class="n">plt</span><span class="o">.</span><span class="n">rc_context</span><span class="p">():</span>  <span class="c1"># Use this to set figure params like size and dpi</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">highest_expr_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">n_top</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">loc</span><span class="o">+</span><span class="s2">&quot;figures/heg.pdf&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/scripts_3_analysis_scripts_6_perturbseq_tsne_kldivergence_umap_4_0.png" src="../../_images/scripts_3_analysis_scripts_6_perturbseq_tsne_kldivergence_umap_4_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/scripts_3_analysis_scripts_6_perturbseq_tsne_kldivergence_umap_4_1.png" src="../../_images/scripts_3_analysis_scripts_6_perturbseq_tsne_kldivergence_umap_4_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">##figure2</span>
<span class="c1">#change plot figure</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.figsize&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">]</span>
<span class="c1">#make prettier</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;white&quot;</span><span class="p">)</span>
<span class="c1">#check if AP and MP present for each gene</span>
<span class="n">value_counts_p1p2</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="s2">&quot;perturbation&quot;</span><span class="p">,</span><span class="s2">&quot;promoter_type&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="n">value_counts_p1p2</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span><span class="n">value_counts_p1p2</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="s1">&#39;perturbation&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s2">&quot;promoter_type&quot;</span><span class="p">,</span><span class="n">values</span><span class="o">=</span><span class="s2">&quot;count&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="n">sns_plot</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">jointplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">value_counts_p1p2</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;AP&quot;</span><span class="p">,</span><span class="n">y</span><span class="o">=</span><span class="s2">&quot;MP&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color5</span><span class="p">,</span>
    <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">marginal_kws</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
<span class="p">)</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">value_counts_p1p2</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
    <span class="c1">#randomn number between 0 and 1</span>
    <span class="n">random_1</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">x_coord</span><span class="o">=</span><span class="n">value_counts_p1p2</span><span class="o">.</span><span class="n">AP</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">+</span><span class="n">random_1</span>
    <span class="n">y_coord</span><span class="o">=</span><span class="n">value_counts_p1p2</span><span class="o">.</span><span class="n">MP</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">+</span><span class="n">random_1</span>
    <span class="n">name</span><span class="o">=</span><span class="n">value_counts_p1p2</span><span class="o">.</span><span class="n">perturbation</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">x_coord</span><span class="p">,</span><span class="n">y</span><span class="o">=</span><span class="n">y_coord</span><span class="p">,</span><span class="n">s</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">value_counts_p1p2</span><span class="o">.</span><span class="n">perturbation</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;ESR1&quot;</span><span class="p">:</span>
        <span class="c1">#style bold</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">value_counts_p1p2</span><span class="o">.</span><span class="n">AP</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">+</span><span class="mf">0.3</span><span class="p">,</span><span class="n">y</span><span class="o">=</span><span class="n">value_counts_p1p2</span><span class="o">.</span><span class="n">MP</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">+</span><span class="mf">0.3</span><span class="p">,</span><span class="n">s</span><span class="o">=</span><span class="n">value_counts_p1p2</span><span class="o">.</span><span class="n">perturbation</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="n">color3</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Number of Cells with </span><span class="se">\n</span><span class="s2"> the P2 Guide For a Gene Present&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Number of Cells with  </span><span class="se">\n</span><span class="s2"> the P1 Guide For a Gene Present&quot;</span><span class="p">)</span>
<span class="c1">#save plot in pdf form</span>
<span class="n">output_fig_loc</span><span class="o">=</span><span class="n">loc</span><span class="o">+</span><span class="s2">&quot;figures/4_5_quality_control/&quot;</span>
<span class="n">sns_plot</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_fig_loc</span><span class="o">+</span><span class="s2">&quot;cellcount_perprom.pdf&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
WARNING:matplotlib.text:posx and posy should be finite values
WARNING:matplotlib.text:posx and posy should be finite values
WARNING:matplotlib.text:posx and posy should be finite values
WARNING:matplotlib.text:posx and posy should be finite values
WARNING:matplotlib.text:posx and posy should be finite values
WARNING:matplotlib.text:posx and posy should be finite values
WARNING:matplotlib.text:posx and posy should be finite values
WARNING:matplotlib.text:posx and posy should be finite values
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/scripts_3_analysis_scripts_6_perturbseq_tsne_kldivergence_umap_5_1.png" src="../../_images/scripts_3_analysis_scripts_6_perturbseq_tsne_kldivergence_umap_5_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">loc</span><span class="o">=</span><span class="s2">&quot;alt-prom-crispr-fiveprime/&quot;</span>
<span class="c1"># specify parameters and distributions to sample from</span>
<span class="c1"># genelist=list(pop.cells[&quot;perturbation&quot;][(pop.cells[&quot;perturbation&quot;]!=&quot;control&quot;) &amp; (pop.cells[&quot;perturbation&quot;]!=&quot;NegCtrl3b&quot;)].drop_duplicates())</span>
<span class="n">genelist</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;perturbation&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">values</span>
<span class="c1">#remove non-targeting</span>
<span class="n">genelist</span><span class="o">=</span><span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">genelist</span> <span class="k">if</span> <span class="n">x</span> <span class="o">!=</span> <span class="s2">&quot;non-targeting&quot;</span><span class="p">]</span>
<span class="n">hyperparam_list</span><span class="o">=</span><span class="p">[]</span>
<span class="n">best_list_list</span><span class="o">=</span><span class="p">[]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#calculate the highly variable genes</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">1500</span><span class="p">,</span> <span class="n">flavor</span><span class="o">=</span><span class="s2">&quot;seurat_v3&quot;</span><span class="p">,</span> <span class="n">n_bins</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/helenking/anaconda3/envs/apu/lib/python3.12/site-packages/scanpy/preprocessing/_highly_variable_genes.py:75: UserWarning: `flavor=&#39;seurat_v3&#39;` expects raw count data, but non-integers were found.
  warnings.warn(
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">highlyvariable_genes</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">[(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;highly_variable&quot;</span><span class="p">]</span><span class="o">==</span><span class="kc">True</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">genelist</span><span class="p">))]</span>
<span class="c1">#add highlyvariable_genes to the targetggene list</span>
<span class="n">adata_sub</span><span class="o">=</span><span class="n">adata</span>
<span class="n">adata_sub</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="n">adata_sub</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_barcode&quot;</span><span class="p">]</span>
<span class="c1">#subset the highly variable genes</span>
<span class="n">adata_sub</span><span class="o">=</span><span class="n">adata</span><span class="p">[:,</span><span class="n">highlyvariable_genes</span><span class="p">]</span>
<span class="n">new_popmatrix_no_duplicates</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata_sub</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;log1p&quot;</span><span class="p">])</span>
<span class="c1"># new_popmatrix_no_duplicates=pd.DataFrame(adata_sub.layers[&quot;sctransform&quot;].todense())</span>
<span class="c1">#columns are genes, rows are cells</span>
<span class="n">new_popmatrix_no_duplicates</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="n">adata_sub</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span>
<span class="n">new_popmatrix_no_duplicates</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="n">adata_sub</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span>
<span class="n">new_popmatrix_no_duplicates_read_count</span><span class="o">=</span><span class="n">new_popmatrix_no_duplicates</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%capture</span>
<span class="c1">#map a umap, calculate kl diveergence, and rand score</span>
<span class="c1"># genelist=[&quot;MYBBP1A&quot;,&quot;ESR1&quot;]</span>
<span class="c1"># genelist=[&quot;MYBBP1A&quot;,&quot;ESR1&quot;, &quot;PSMC5&quot;,&quot;BRIP1&quot;,&quot;STAT3&quot;,&quot;CHD4&quot;]</span>
<span class="c1">#change the backgorun dto white</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genelist</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="c1">##list_guides from adata</span>
    <span class="n">list_guides_df</span><span class="o">=</span><span class="n">adata_sub</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">adata_sub</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;perturbation&quot;</span><span class="p">]</span><span class="o">==</span><span class="n">gene</span><span class="p">]</span>
    <span class="n">list_guides</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">list_guides_df</span><span class="p">[</span><span class="s2">&quot;guide_identity&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">())</span>
    <span class="c1">#find the gene name of interest</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">list_guides</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;skip&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">new_popmatrix_no_duplicates</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;skip&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1">#subset cells+in</span>
        <span class="c1"># cells_interest=new_popmatrix_no_duplicates[new_popmatrix_no_duplicates.index.isin(list_guides_df[&quot;cell_barcode&quot;].drop_duplicates())]</span>
        <span class="n">cells_interest</span><span class="o">=</span><span class="n">adata_sub</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">adata_sub</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">list_guides_df</span><span class="p">[</span><span class="s2">&quot;cell_barcode&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">())]</span>
        <span class="c1"># cells_interest[&quot;cell_barcode&quot;]=cells_interest.index</span>
        <span class="n">cells_interest</span><span class="o">=</span><span class="n">cells_interest</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s2">&quot;cell_barcode&quot;</span><span class="p">)</span>
        <span class="n">column_nopromotergene</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">new_popmatrix_no_duplicates</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="p">((</span><span class="s1">&#39;MT&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">col</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="s1">&#39;Non&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">col</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="s1">&#39;sg&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">col</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="s1">&#39;_&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">col</span><span class="p">))]</span>
        <span class="nb">all</span><span class="o">=</span><span class="n">new_popmatrix_no_duplicates</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">cells_interest</span><span class="p">[</span><span class="s2">&quot;cell_barcode&quot;</span><span class="p">],</span><span class="n">column_nopromotergene</span><span class="p">]</span>
        <span class="c1">#this creates one list that contains repeating elements one for</span>
        <span class="n">list_rep</span><span class="o">=</span><span class="n">cells_interest</span><span class="p">[</span><span class="s2">&quot;promoter_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="c1">##maybe make an if loop that if one of the types are not present then skip....</span>
        <span class="n">list_rep</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">squeeze</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">list_rep</span><span class="p">))</span>
        <span class="c1"># list_rep_cellcycle=cells_interest[&quot;phase&quot;].values</span>
        <span class="c1"># list_rep_cellcycle=np.squeeze(pd.DataFrame(list_rep_cellcycle))</span>
        <span class="n">standard_embedding</span> <span class="o">=</span> <span class="n">umap</span><span class="o">.</span><span class="n">UMAP</span><span class="p">(</span><span class="n">random_state</span><span class="o">=</span><span class="mi">22</span><span class="p">)</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="nb">all</span><span class="p">)</span>
        <span class="c1">##hyperparam</span>
        <span class="n">hyper_paramlist</span><span class="o">=</span><span class="n">hyperparam</span><span class="p">(</span><span class="n">df</span> <span class="o">=</span> <span class="n">standard_embedding</span><span class="p">,</span> <span class="n">list_rep</span><span class="o">=</span><span class="n">list_rep</span><span class="p">)</span>
        <span class="n">best</span> <span class="o">=</span> <span class="n">best_validity</span><span class="p">(</span><span class="n">hyper_paramlist</span><span class="p">)</span>
        <span class="n">clust_alg_eom</span> <span class="o">=</span> <span class="n">hdbscan</span><span class="o">.</span><span class="n">HDBSCAN</span><span class="p">(</span><span class="n">algorithm</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
                                    <span class="n">approx_min_span_tree</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                    <span class="n">gen_min_span_tree</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                    <span class="n">cluster_selection_method</span><span class="o">=</span><span class="n">best</span><span class="p">[</span><span class="s1">&#39;method&#39;</span><span class="p">],</span>
                                    <span class="n">metric</span><span class="o">=</span><span class="n">best</span><span class="p">[</span><span class="s1">&#39;metric&#39;</span><span class="p">],</span>
                                    <span class="n">min_cluster_size</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">best</span><span class="p">[</span><span class="s1">&#39;min_cluster_size&#39;</span><span class="p">]),</span>
                                    <span class="n">min_samples</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">best</span><span class="p">[</span><span class="s1">&#39;min_samples&#39;</span><span class="p">]),</span>
                                    <span class="n">allow_single_cluster</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">standard_embedding</span><span class="p">)</span>
        <span class="n">best_list</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">best</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
        <span class="n">best_list</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">gene</span>
        <span class="n">best_list_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">best_list</span><span class="p">)</span>
        <span class="c1">##all</span>
        <span class="n">df_list</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">hyper_paramlist</span><span class="p">)</span>
        <span class="n">df_list</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">gene</span>
        <span class="n">hyperparam_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">df_list</span><span class="p">)</span>

        <span class="c1">###!!!plotting section !!!!####</span>
        <span class="c1">##individual plotting</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span>  <span class="n">dpi</span><span class="o">=</span> <span class="mi">600</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">)</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;white&quot;</span><span class="p">,</span><span class="n">font_scale</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span><span class="n">rc</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;figure.figsize&#39;</span><span class="p">:(</span><span class="mi">4</span><span class="p">,</span><span class="mi">3</span><span class="p">)})</span>
        <span class="n">sns_plot</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">standard_embedding</span><span class="p">[:,</span><span class="mi">1</span><span class="p">],</span><span class="n">y</span><span class="o">=</span><span class="n">standard_embedding</span><span class="p">[:,</span><span class="mi">0</span><span class="p">],</span>  <span class="n">hue</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">list_rep</span><span class="o">==</span><span class="s2">&quot;MP&quot;</span><span class="p">,</span> <span class="s2">&quot;P1&quot;</span><span class="p">,</span> <span class="s2">&quot;P2&quot;</span><span class="p">),</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span><span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>

        <span class="c1">###with cell cycle assignment</span>
        <span class="c1"># sns_plot=sns.scatterplot(x=standard_embedding[:,1],y=standard_embedding[:,0],  hue=np.where(list_rep==&quot;MP&quot;, &quot;P1&quot;, &quot;P2&quot;),ax=ax[0],palette=palette)</span>
        <span class="c1"># sns_plot=sns.scatterplot(x=standard_embedding[:,1],y=standard_embedding[:,0],  hue=list_rep_cellcycle,ax=ax[1],palette=cellcycle_palette)</span>
        <span class="c1">#place the legend outside</span>
        <span class="c1"># ax[0].get_legend().remove()</span>
        <span class="c1"># ax[0].legend(loc=&#39;center left&#39;, bbox_to_anchor=(1, 0.5))</span>

        <span class="c1">#tsne4.kl_divergence_</span>
        <span class="c1">###!!!cluster!!!####</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">best</span><span class="p">)</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">sns_plot</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">standard_embedding</span><span class="p">[:,</span><span class="mi">1</span><span class="p">],</span><span class="n">y</span><span class="o">=</span><span class="n">standard_embedding</span><span class="p">[:,</span><span class="mi">0</span><span class="p">],</span>  <span class="n">hue</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">list_rep</span><span class="o">==</span><span class="s2">&quot;MP&quot;</span><span class="p">,</span> <span class="s2">&quot;P1&quot;</span><span class="p">,</span> <span class="s2">&quot;P2&quot;</span><span class="p">),</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">umap_palette</span><span class="p">)</span>
        <span class="n">sns_plot</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">standard_embedding</span><span class="p">[:,</span><span class="mi">1</span><span class="p">],</span><span class="n">y</span><span class="o">=</span><span class="n">standard_embedding</span><span class="p">[:,</span><span class="mi">0</span><span class="p">],</span>  <span class="n">hue</span><span class="o">=</span><span class="n">clust_alg_eom</span><span class="o">.</span><span class="n">labels_</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">umap_palette</span><span class="p">)</span>
        <span class="c1">#</span>
        <span class="n">sns_plot</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;alt-prom-crispr-fiveprime/figures/umap/&quot;</span><span class="o">+</span><span class="n">gene</span><span class="o">+</span><span class="s2">&quot;.pdf&quot;</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#hyperparam</span>
<span class="n">dataframe</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">hyperparam_list</span><span class="p">)</span>
<span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;metric&#39;</span><span class="p">,</span><span class="s1">&#39;min_cluster_size&#39;</span><span class="p">,</span> <span class="s1">&#39;method&#39;</span><span class="p">,</span><span class="s1">&#39;min_samples&#39;</span><span class="p">,</span> <span class="s1">&#39;validity_score&#39;</span><span class="p">,</span> <span class="s1">&#39;n_clusters&#39;</span><span class="p">,</span><span class="s1">&#39;randscore&#39;</span><span class="p">,</span><span class="s1">&#39;mutualscore&#39;</span> <span class="p">,</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span>
<span class="n">dataframe</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="n">cols</span>
<span class="n">dataframe</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;alt-prom-crispr-fiveprime/files/singlecell_shortread_analysis/hyperparam_cluster.csv&quot;</span><span class="p">)</span>

<span class="c1">#metric min_cluster_size method min_samples validity_score n_clusters randscore mutualscore   gene</span>
<span class="n">dataframe</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">best_list_list</span><span class="p">)</span>
<span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;metric&#39;</span><span class="p">,</span><span class="s1">&#39;min_cluster_size&#39;</span><span class="p">,</span> <span class="s1">&#39;method&#39;</span><span class="p">,</span><span class="s1">&#39;min_samples&#39;</span><span class="p">,</span> <span class="s1">&#39;validity_score&#39;</span><span class="p">,</span> <span class="s1">&#39;n_clusters&#39;</span><span class="p">,</span><span class="s1">&#39;randscore&#39;</span><span class="p">,</span><span class="s1">&#39;mutualscore&#39;</span><span class="p">,</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span>
<span class="n">dataframe</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="n">cols</span>
<span class="n">dataframe</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;alt-prom-crispr-fiveprime/files/singlecell_shortread_analysis/best_cluster.csv&quot;</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hyperparam_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;alt-prom-crispr-fiveprime/files/singlecell_shortread_analysis/hyperparam_cluster.csv&quot;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">best_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;alt-prom-crispr-fiveprime/files/singlecell_shortread_analysis/best_cluster.csv&quot;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">best_df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Quantifying Separation with Rand Index and Mutual Information To move beyond visual interpretation, the notebook uses rigorous statistical metrics to quantify how “different” the P1 and P2 populations really are.</p>
<p>Rand Index (Adjusted Rand Score): This measures the similarity between the “ground truth” (which promoter was targeted) and unsupervised clusters. A higher score confirms that the transcriptomic differences are strong enough for an algorithm to “guess” the targeted promoter.</p>
<p>Mutual Information (MI): This information-theory metric quantifies the amount of information obtained about the promoter perturbation by observing the cell’s transcriptome.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#plot the validity_score randscore and mutualscore</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;white&quot;</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="s2">&quot;ticks&quot;</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_context</span><span class="p">(</span><span class="s2">&quot;paper&quot;</span><span class="p">,</span> <span class="n">font_scale</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">best_df</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;mutualscore&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;randscore&quot;</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>
<span class="c1">#label the poiints abopve 0.04 mutual score</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">best_df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
    <span class="k">if</span> <span class="n">best_df</span><span class="o">.</span><span class="n">mutualscore</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">&gt;</span><span class="mf">0.04</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">best_df</span><span class="o">.</span><span class="n">mutualscore</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="n">best_df</span><span class="o">.</span><span class="n">randscore</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="n">best_df</span><span class="o">.</span><span class="n">gene</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="c1">#label the two</span>

<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Mutual Information Score&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Rand Score&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;alt-prom-crispr-fiveprime/figures/umap/mutual_rand.pdf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/scripts_3_analysis_scripts_6_perturbseq_tsne_kldivergence_umap_13_0.png" src="../../_images/scripts_3_analysis_scripts_6_perturbseq_tsne_kldivergence_umap_13_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[75]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;successfulKD&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;True&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[75]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
cell_barcode
GTGTGCGGTTAGTGGG-1       NBN
TCAGATGAGCATCATC-1     PA2G4
GCAATCACACGACGAA-1       PKM
CACACAAAGCTGCAAG-1     SIN3A
GTCACGGGTTAAGTAG-1       SP1
                       ...
TAGCCGGAGCGTAATA-1     NCOR2
CTTGGCTTCAGCTTAG-1      CHD4
CATATGGAGGTGTTAA-1     GPBP1
TCTCTAAAGGAGCGTT-1     KDM2A
GGGAATGTCTTGACGA-1    CTNNB1
Name: gene, Length: 21657, dtype: category
Categories (50, object): [&#39;ADAR&#39;, &#39;AHCYL1&#39;, &#39;APP&#39;, &#39;ARNT&#39;, ..., &#39;XRCC5&#39;, &#39;YWHAZ&#39;, &#39;ZNF3&#39;, &#39;ZNF217&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[79]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">vis_gene</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;MYBBP1A&quot;</span><span class="p">,</span><span class="s2">&quot;ESR1&quot;</span><span class="p">,</span> <span class="s2">&quot;GTF2F1&quot;</span><span class="p">,</span><span class="s2">&quot;BRIP1&quot;</span><span class="p">,</span><span class="s2">&quot;STAT3&quot;</span><span class="p">,</span><span class="s2">&quot;CHD4&quot;</span><span class="p">]</span>
<span class="n">subset_suc</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;successfulKD&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;True&quot;</span><span class="p">]</span>

<span class="n">subset_suc</span><span class="p">[</span><span class="n">subset_suc</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">vis_gene</span><span class="p">)]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[79]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;CHD4&#39;, &#39;ESR1&#39;, &#39;MYBBP1A&#39;, &#39;STAT3&#39;, &#39;BRIP1&#39;, &#39;GTF2F1&#39;]
Categories (50, object): [&#39;ADAR&#39;, &#39;AHCYL1&#39;, &#39;APP&#39;, &#39;ARNT&#39;, ..., &#39;XRCC5&#39;, &#39;YWHAZ&#39;, &#39;ZNF3&#39;, &#39;ZNF217&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#read in edistance and check the correlation per gene</span>
<span class="n">edistance</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;alt-prom-crispr-fiveprime/files/singlecell_shortread_analysis/edistance.csv&quot;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">edistance</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">edistance</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">edistance</span><span class="p">[</span><span class="s2">&quot;promoter&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">edistance</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="c1">#merge edistance and best_df</span>
<span class="n">edistance</span><span class="o">=</span><span class="n">edistance</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">best_df</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>
<span class="c1">#check correlation between edistance and randscore</span>
<span class="c1">#perform spearmans correlation between the two</span>
<span class="n">edistance</span><span class="p">[</span><span class="s2">&quot;edist&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">edistance</span><span class="p">[</span><span class="s2">&quot;edist&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
<span class="n">edistance</span><span class="p">[</span><span class="s2">&quot;randscore&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">edistance</span><span class="p">[</span><span class="s2">&quot;randscore&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">font_scale</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># crazy big</span>

<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">edistance</span><span class="p">[[</span><span class="s1">&#39;edist&#39;</span><span class="p">,</span><span class="s1">&#39;randscore&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;pearson&quot;</span><span class="p">),</span> <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;crest&quot;</span><span class="p">,</span> <span class="n">linewidths</span><span class="o">=</span><span class="mf">.5</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cbar_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;shrink&quot;</span><span class="p">:</span> <span class="mf">.5</span><span class="p">})</span>
<span class="c1">#change the size of text</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;alt-prom-crispr-fiveprime/figures/edistance_randscore.pdf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/scripts_3_analysis_scripts_6_perturbseq_tsne_kldivergence_umap_16_0.png" src="../../_images/scripts_3_analysis_scripts_6_perturbseq_tsne_kldivergence_umap_16_0.png" />
</div>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="5_genekd_neighbouring_gene_expression.html" class="btn btn-neutral float-left" title="Target gene and neighbouring gene expression" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="7_cellphase_model.html" class="btn btn-neutral float-right" title="Cell Assignment Model from FUCCI-Matched Single-Cell Paper" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2026, Helen King.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>