

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Cell Assignment Model from FUCCI-Matched Single-Cell Paper &mdash; Isoform Specific Perturb-Seq  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=9edc463e" />
      <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
      <link rel="stylesheet" type="text/css" href="../../_static/nbsphinx-code-cells.css?v=2aa19091" />

  
      <script src="../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../../_static/doctools.js?v=fd6eb6e6"></script>
      <script src="../../_static/sphinx_highlight.js?v=6ffebe34"></script>
      <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
      <script src="../../_static/copybutton.js?v=f281be69"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@4/tex-mml-chtml.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Differential Expression analysis between P1 &amp; P2" href="8_perturbseq_differentialexpression.html" />
    <link rel="prev" title="Quantitative individual UMAPs per gene of interest" href="6_perturbseq_tsne_kldivergence_umap.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            Isoform Specific Perturb-Seq
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Analysis Pipeline:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../promoter_identification.html">Promoter Identification</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../guide_design.html">Guide Design</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../analysis_scripts.html">Analysis Scripts &amp; Pipelines</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../analysis_scripts.html#preprocessing-guide-calling">Preprocessing &amp; Guide Calling</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../analysis_scripts.html#splicing-rna-velocity">Splicing &amp; RNA Velocity</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../analysis_scripts.html#transcriptional-divergence-e-stats">Transcriptional Divergence (E-Stats)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../analysis_scripts.html#cell-phase-pathway-modeling">Cell Phase &amp; Pathway Modeling</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../analysis_scripts.html#clinical-differential-expression">Clinical &amp; Differential Expression</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="2A_cellranger_guidecalling.html">Guide Calling for Dual Guides</a></li>
<li class="toctree-l3"><a class="reference internal" href="2B_guide_calling_cellrangerfilter4_idealguides.html">Filtering for Cells with Ideal Combination of Guides</a></li>
<li class="toctree-l3"><a class="reference internal" href="4_estatistic_gene_guide.html">Guide Calling Using the Cellranger Module HDF5 File</a></li>
<li class="toctree-l3"><a class="reference internal" href="5_genekd_neighbouring_gene_expression.html">Target gene and neighbouring gene expression</a></li>
<li class="toctree-l3"><a class="reference internal" href="6_perturbseq_tsne_kldivergence_umap.html">Quantitative individual UMAPs per gene of interest</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Cell Assignment Model from FUCCI-Matched Single-Cell Paper</a></li>
<li class="toctree-l3"><a class="reference internal" href="8_perturbseq_differentialexpression.html">Differential Expression analysis between P1 &amp; P2</a></li>
<li class="toctree-l3"><a class="reference internal" href="9A_cnv_score.html">Investigating Copy Number Variation Across KD Populations</a></li>
<li class="toctree-l3"><a class="reference internal" href="9B_velocity_loom.html">Velocity and pyGSEA calculations of ESR1</a></li>
<li class="toctree-l3"><a class="reference internal" href="11_spectra.html">Spectra Analysis</a></li>
<li class="toctree-l3"><a class="reference internal" href="12_negativecontrol.html">Guide Calling Using the Cellranger Module HDF5 File</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Isoform Specific Perturb-Seq</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../analysis_scripts.html">Analysis Scripts &amp; Pipelines</a></li>
      <li class="breadcrumb-item active">Cell Assignment Model from FUCCI-Matched Single-Cell Paper</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/scripts/3_analysis_scripts/7_cellphase_model.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Cell-Assignment-Model-from-FUCCI-Matched-Single-Cell-Paper">
<h1>Cell Assignment Model from FUCCI-Matched Single-Cell Paper<a class="headerlink" href="#Cell-Assignment-Model-from-FUCCI-Matched-Single-Cell-Paper" title="Link to this heading"></a></h1>
<ol class="arabic simple">
<li><p>Run through the 10x Cellranger pipeline and velocyto for single cell RNAseq quatification and using (2) guides quantifiction. all found in the cellranger files folder <strong>bash</strong></p></li>
<li><p>Guide Calling for dual guide. Use repogle method to take molecule.h5 generated by cellranger and py to run through repogle version of guide calling or use cellranger_guidecalling.ipynb for Direct Capture Perturb-Seq dual guide. Formed guide-specific lists of cells.</p></li>
<li><p>Pseudobulk analysis. A. Seperation of guide-specific fastq files. <strong>bash</strong> B. Whippet pseudobulk for transcript specific analysis, post UMI deduplication. <strong>bash</strong> C. Transcript quality control. <strong>R</strong> D. Whippet result visualisation.</p></li>
<li><p>Normalisation of adata object and E-distance of KD</p></li>
<li><p>Check gene and neighboring gene expression</p></li>
<li><p>Create individual umaps per gene of interest A. UMAPs B. Rand Index score</p></li>
<li><p><strong>Cell phase assignment model from FUCCI-matched single cell paper (GSE146773)</strong></p></li>
<li><p>Differential Expression analysis. A. Find the shared P1 and P2 genes. B. Check the shared P1 and P2 across different protospacers with the same A/B and C/D.</p></li>
<li><p>CNV Score &amp; Numbat to quantify and Velocity quantification with loom file</p></li>
<li><p>ESR1-specific analysis from proliferation analysis to rt-qpcr</p></li>
<li><p>Spectra analysis and visualisation for pathway enrichment</p></li>
</ol>
<p>This script is dedicated to the functional phenotypic validation of the Isoform-Specific single-cell Perturb-Seq data. Specifically, it implements a machine learning pipeline to assign cell cycle phases (G1, S, and G2/M) to every individual cell in the screen to determine how specific promoter knockdowns disrupt cellular proliferation.</p>
<p>Training a Robust Cell Cycle Classifier Instead of relying on generic markers, the authors trained a specialized model to accurately identify cell states.</p>
<p>Reference Dataset: The model uses a public dataset (GSE146773) where cells were physically sorted by FUCCI (Fluorescent Ubiquitination-based Cell Cycle Indicator) and then sequenced, providing a “ground truth” for the cell cycle phase.</p>
<p>K-Nearest Neighbors (KNN): The code trains a KNN classifier on this reference data using highly variable cell cycle-regulated genes.</p>
<p>Performance Validation: The notebook uses ROC Curves to verify accuracy, achieving a score of 0.84, which significantly outperforms the standard cell cycle scoring methods typically used in single-cell analysis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">load_ext</span> autoreload
<span class="o">%</span><span class="k">matplotlib</span> inline
<span class="o">%</span><span class="k">autoreload</span> 2
<span class="c1">#general</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pl</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ad</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">hdf5plugin</span>

<span class="c1">#form a location</span>
<span class="n">loc</span><span class="o">=</span><span class="s2">&quot;alt-prom-crispr-fiveprime/&quot;</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">tqdm.notebook</span><span class="w"> </span><span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">loc</span><span class="o">+</span><span class="s1">&#39;scripts/&#39;</span><span class="p">)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">apu_analysis</span><span class="w"> </span><span class="kn">import</span> <span class="o">*</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scperturb</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">infercnvpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">cnv</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">apu_analysis.cell_import</span><span class="w"> </span><span class="kn">import</span> <span class="n">CellPopulation</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">IPython.display</span><span class="w"> </span><span class="kn">import</span> <span class="n">clear_output</span>
<span class="n">pd</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">display</span><span class="o">.</span><span class="n">float_format</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{:.4f}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">statsmodels.stats.multitest</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">multi</span>

<span class="c1">#for this python</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.special</span><span class="w"> </span><span class="kn">import</span> <span class="n">rel_entr</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sklearn.cluster</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">cluster</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">umap</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy</span><span class="w"> </span><span class="kn">import</span> <span class="n">stats</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.stats</span><span class="w"> </span><span class="kn">import</span> <span class="n">bootstrap</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">statsmodels.api</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sm</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">statsmodels.stats.multitest</span><span class="w"> </span><span class="kn">import</span> <span class="n">multipletests</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="kn">import</span> <span class="n">reshape</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="kn">import</span> <span class="n">array</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.decomposition</span><span class="w"> </span><span class="kn">import</span> <span class="n">PCA</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">hdbscan</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">statsmodels.stats.weightstats</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ws</span>

<span class="c1"># Taken from:</span>
<span class="c1"># Adamson, B.A., Norman, T.M., *et al.* &quot;A multiplexed CRISPR screening platform enables systematic dissection of the unfolded protein response&quot;, *Cell*, 2016.</span>
<span class="c1"># My experiment deals with two KDs- one of the MP, one of the AP using two guides. Positve controls include GINS1 ect. This is a combnatorial KD double for the same gene. No treatments were used</span>

<span class="c1"># colours using garvan</span>
<span class="n">color1</span> <span class="o">=</span><span class="s1">&#39;#4d00c7&#39;</span>
<span class="n">palecolor1</span><span class="o">=</span><span class="s2">&quot;#b366ff&quot;</span>
<span class="n">color2</span><span class="o">=</span> <span class="s1">&#39;#da3c07&#39;</span>
<span class="n">palecolor2</span><span class="o">=</span><span class="s2">&quot;#ff8954&quot;</span>
<span class="n">color3</span><span class="o">=</span><span class="s1">&#39;#05d3d3&#39;</span>
<span class="n">color4</span><span class="o">=</span><span class="s1">&#39;#c6c7c5&#39;</span>
<span class="n">color4</span><span class="o">=</span><span class="s2">&quot;#434541&quot;</span>
<span class="n">color5</span><span class="o">=</span><span class="s2">&quot;#eb31e1&quot;</span>
<span class="n">color6</span><span class="o">=</span><span class="s2">&quot;#3175eb&quot;</span>
<span class="n">color7</span><span class="o">=</span><span class="s2">&quot;#a7eb31&quot;</span>
<span class="n">color8</span><span class="o">=</span><span class="s2">&quot;#b366ff&quot;</span>
<span class="n">color9</span><span class="o">=</span><span class="s2">&quot;#ff8954&quot;</span>
<span class="n">color10</span><span class="o">=</span><span class="s2">&quot;#35c9d4&quot;</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>

<span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s1">&#39;OMP_NUM_THREADS&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;1&#39;</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scipy.stats</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">stats</span>
<span class="c1">###guide_num</span>

<span class="c1"># Create the color palette</span>
<span class="n">palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">([</span><span class="n">palecolor1</span><span class="p">,</span><span class="n">palecolor2</span><span class="p">])</span>
<span class="n">palette2</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">([</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span> <span class="n">color3</span><span class="p">,</span> <span class="n">color4</span><span class="p">,</span><span class="n">color5</span><span class="p">,</span><span class="n">color6</span> <span class="p">,</span><span class="n">color7</span><span class="p">])</span>

<span class="c1"># Create the color palette</span>
<span class="n">palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">([</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span><span class="n">color3</span><span class="p">])</span>
<span class="n">new_palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">([</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span> <span class="n">color3</span><span class="p">,</span> <span class="n">color4</span><span class="p">])</span>

<span class="c1">#change palette to 3d82c4 36a047</span>
<span class="n">palette</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">([</span><span class="s2">&quot;#f3766e&quot;</span><span class="p">,</span> <span class="s2">&quot;#7094cd&quot;</span><span class="p">])</span>
<span class="n">cellcycle_palette</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">([</span><span class="s2">&quot;#36a047&quot;</span><span class="p">,</span> <span class="s2">&quot;#7671b3&quot;</span><span class="p">,</span><span class="s2">&quot;#d76127&quot;</span><span class="p">])</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Scanpy&quot;</span><span class="p">,</span> <span class="n">sc</span><span class="o">.</span><span class="n">__version__</span><span class="p">)</span>
<span class="o">%</span><span class="k">matplotlib</span> inline
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload
Scanpy 1.10.3
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="c1">#onvert the ensembl_id tohgnc_gene_id</span>
<span class="c1">#get the ensembl_gene_id for each hgnc_symbol</span>
<span class="n">annot</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">queries</span><span class="o">.</span><span class="n">biomart_annotations</span><span class="p">(</span>
            <span class="s2">&quot;hsapiens&quot;</span><span class="p">,</span>
            <span class="p">[</span><span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">,</span><span class="s2">&quot;hgnc_symbol&quot;</span><span class="p">,</span><span class="s2">&quot;ensembl_peptide_id&quot;</span> <span class="p">],</span>
        <span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;hgnc_symbol&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
<br/><br/><br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">loc</span><span class="o">+</span><span class="s2">&quot;files/adata_normalised.h5ad&quot;</span><span class="p">)</span>
<span class="c1">#subset for successful knockdowns</span>
<span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">[(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;successfulKD&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;True&quot;</span> <span class="p">)</span><span class="o">|</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;non-targeting_Control&quot;</span><span class="p">)</span> <span class="p">]</span>
<span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="c1">#make sure log1p</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="c1">#filter adata with highly variable genes</span>
<span class="c1"># adata=adata[:,adata.var[&quot;highly_variable&quot;]==True]</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">annot</span><span class="p">[[</span><span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(),</span> <span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;gene_id&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="c1">#find ensembl_id</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/helenking/anaconda3/envs/apu/lib/python3.12/site-packages/anndata/_core/anndata.py:618: FutureWarning: You are attempting to set `X` to a matrix on a view which has non-unique indices. The resulting `adata.X` will likely not equal the value to which you set it. To avoid this potential issue, please make a copy of the data first. In the future, this operation will throw an error.
  warnings.warn(msg, FutureWarning, stacklevel=1)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: adata.X seems to be already log-transformed.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/helenking/anaconda3/envs/apu/lib/python3.12/site-packages/scanpy/preprocessing/_simple.py:406: UserWarning: Received a view of an AnnData. Making a copy.
  view_to_actual(adata)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#provide the counts and labels for trianing data</span>
<span class="n">loc</span><span class="o">=</span><span class="s2">&quot;alt-prom-crispr-fiveprime/&quot;</span>
<span class="n">counts_training_data</span><span class="o">=</span><span class="n">loc</span><span class="o">+</span><span class="s2">&quot;files/cellcycle_assignment/GSE146773_Counts.csv.gz&quot;</span>
<span class="n">labels_training_data</span><span class="o">=</span><span class="n">loc</span><span class="o">+</span><span class="s2">&quot;files/cellcycle_assignment/180911_*csv&quot;</span>
<span class="n">human_genes_loc</span><span class="o">=</span><span class="n">loc</span><span class="o">+</span><span class="s2">&quot;files/cellcycle_assignment/human_periodic.tsv&quot;</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#form adata dataframe</span>
<span class="n">GSE146773_adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">counts_training_data</span><span class="p">)</span>
<span class="c1">#filter the genes get the</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">GSE146773_adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">GSE146773_adata</span><span class="p">)</span>
<span class="c1">#convert the adata.</span>
<span class="n">GSE146773_adata</span><span class="o">.</span><span class="n">var</span><span class="o">=</span><span class="n">GSE146773_adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">annot</span><span class="p">[[</span><span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(),</span> <span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
<span class="c1">#remove na in GSE146773_adata.var</span>
<span class="c1">#make var_names a string</span>
<span class="n">GSE146773_adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">=</span><span class="n">GSE146773_adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
<span class="n">GSE146773_adata</span><span class="o">=</span><span class="n">GSE146773_adata</span><span class="p">[:,</span><span class="o">~</span><span class="n">GSE146773_adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isna</span><span class="p">()]</span>
<span class="n">GSE146773_adata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
<span class="c1">###scanpy</span>
<span class="n">cell_cycle_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">loc</span><span class="o">+</span><span class="s1">&#39;files/cellcycle_assignment/regev_lab_cell_cycle_genes.txt&#39;</span><span class="p">)]</span>
<span class="n">cell_cycle_genes</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">cell_cycle_genes</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">cell_cycle_genes</span><span class="o">.</span><span class="n">shape</span> <span class="p">)</span>
<span class="n">cell_cycle_genes</span><span class="o">=</span><span class="n">cell_cycle_genes</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">annot</span><span class="p">,</span> <span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">cell_cycle_genes</span><span class="o">.</span><span class="n">shape</span> <span class="p">)</span>
<span class="n">cell_cycle_genes</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;hgnc_symbol&quot;</span><span class="p">,</span><span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">,</span><span class="s2">&quot;protein_id&quot;</span><span class="p">]</span>
<span class="n">s_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">cell_cycle_genes</span><span class="p">[</span><span class="s2">&quot;hgnc_symbol&quot;</span><span class="p">][:</span><span class="mi">43</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">GSE146773_adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">]</span>
<span class="n">g2m_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">cell_cycle_genes</span><span class="p">[</span><span class="s2">&quot;hgnc_symbol&quot;</span><span class="p">][</span><span class="mi">43</span><span class="p">:]</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">GSE146773_adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">s_genes</span><span class="p">)</span> <span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">g2m_genes</span><span class="p">))</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(97, 1)
(97, 3)
42 52
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/helenking/anaconda3/envs/apu/lib/python3.12/site-packages/anndata/_core/anndata.py:1756: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  utils.warn_names_duplicates(&#34;var&#34;)
/Users/helenking/anaconda3/envs/apu/lib/python3.12/site-packages/anndata/_core/anndata.py:1756: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  utils.warn_names_duplicates(&#34;var&#34;)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">score_genes_cell_cycle</span><span class="p">(</span><span class="n">GSE146773_adata</span><span class="p">,</span> <span class="n">s_genes</span><span class="o">=</span><span class="n">s_genes</span><span class="p">,</span> <span class="n">g2m_genes</span><span class="o">=</span><span class="n">g2m_genes</span><span class="p">)</span>
<span class="n">GSE146773_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;G1_score&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">GSE146773_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;G2M_score&quot;</span><span class="p">]</span><span class="o">/</span><span class="n">GSE146773_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;S_score&quot;</span><span class="p">]</span>

<span class="c1">#read in the seurat input assignment</span>
<span class="n">GSE146773_counts_labels</span><span class="o">=</span><span class="n">label_from</span><span class="p">(</span><span class="n">labels_training_data</span><span class="p">)</span>
<span class="n">GSE146773_counts_labels</span><span class="o">=</span><span class="n">GSE146773_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">GSE146773_counts_labels</span><span class="p">,</span><span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s2">&quot;cell_name&quot;</span><span class="p">)</span>
<span class="c1">#fill na as S</span>
<span class="n">GSE146773_counts_labels</span><span class="p">[</span><span class="s2">&quot;Population&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">GSE146773_counts_labels</span><span class="p">[</span><span class="s2">&quot;Population&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;S-ph&quot;</span><span class="p">,</span><span class="s2">&quot;S&quot;</span><span class="p">,</span><span class="n">GSE146773_counts_labels</span><span class="p">[</span><span class="s2">&quot;Population&quot;</span><span class="p">])</span>

<span class="n">s_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">cell_cycle_genes</span><span class="p">[</span><span class="s2">&quot;hgnc_symbol&quot;</span><span class="p">][:</span><span class="mi">43</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">]</span>
<span class="n">g2m_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">cell_cycle_genes</span><span class="p">[</span><span class="s2">&quot;hgnc_symbol&quot;</span><span class="p">][</span><span class="mi">43</span><span class="p">:]</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">]</span>
<span class="c1">#add s_genes and s_genes_reg tgethe</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">score_genes_cell_cycle</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">s_genes</span><span class="o">=</span><span class="n">s_genes</span><span class="p">,</span> <span class="n">g2m_genes</span><span class="o">=</span><span class="n">g2m_genes</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="p">(</span><span class="n">knn</span><span class="p">,</span> <span class="n">X_train</span><span class="p">,</span> <span class="n">X_test</span><span class="p">,</span> <span class="n">y_train</span><span class="p">,</span> <span class="n">y_test</span><span class="p">,</span><span class="n">gene_subset</span><span class="p">,</span><span class="n">scaler</span><span class="p">)</span><span class="o">=</span><span class="n">train_model</span><span class="p">(</span><span class="n">counts_training_data</span><span class="p">,</span><span class="n">labels_training_data</span><span class="p">,</span><span class="n">human_genes_loc</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">genes_regev</span><span class="o">=</span><span class="n">cell_cycle_genes</span><span class="p">,</span><span class="n">genelist</span><span class="o">=</span><span class="s2">&quot;regev&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/helenking/anaconda3/envs/apu/lib/python3.12/site-packages/scanpy/preprocessing/_simple.py:406: UserWarning: Received a view of an AnnData. Making a copy.
  view_to_actual(adata)
OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
... storing &#39;Population&#39; as categorical
... storing &#39;Sample&#39; as categorical
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file figures/umapcellcycle_assignment.pdf
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/scripts_3_analysis_scripts_7_cellphase_model_9_2.png" src="../../_images/scripts_3_analysis_scripts_7_cellphase_model_9_2.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Population
G2M    387
G1     346
S      334
Name: count, dtype: int64
Best score: 0.8361272530286614
Best number of neighbors: 5
Best parameters: {&#39;metric&#39;: &#39;euclidean&#39;, &#39;n_neighbors&#39;: 5, &#39;weights&#39;: &#39;uniform&#39;}
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#replace na in list with &quot;S&quot;</span>
<span class="c1"># y_test=[x if x in [&quot;G1&quot;,&quot;G2M&quot;,&quot;S&quot;] else &quot;S&quot; for x in y_test]</span>
<span class="c1">#change S in GSE146773_counts_labels[&quot;Population&quot;] to S-ph</span>
<span class="n">GSE146773_counts_labels</span><span class="p">[</span><span class="s2">&quot;Population&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">GSE146773_counts_labels</span><span class="p">[</span><span class="s2">&quot;Population&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;S-ph&quot;</span><span class="p">,</span><span class="s2">&quot;S&quot;</span><span class="p">,</span><span class="n">GSE146773_counts_labels</span><span class="p">[</span><span class="s2">&quot;Population&quot;</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">((</span><span class="n">gene_subset</span><span class="p">)</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()))</span>
<span class="n">cell_cycle_scores</span><span class="o">=</span><span class="n">compare_seurat</span><span class="p">(</span><span class="n">X_test</span><span class="p">,</span><span class="n">y_test</span><span class="p">,</span><span class="n">GSE146773_counts_labels</span> <span class="p">,</span><span class="n">knn</span><span class="p">,</span> <span class="n">color1</span><span class="o">=</span><span class="n">color1</span><span class="p">,</span><span class="n">color2</span><span class="o">=</span><span class="n">color2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
93
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/scripts_3_analysis_scripts_7_cellphase_model_10_1.png" src="../../_images/scripts_3_analysis_scripts_7_cellphase_model_10_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_cycle_scores</span><span class="o">=</span><span class="n">cell_cycle_scores</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">GSE146773_counts_labels</span><span class="p">,</span> <span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s2">&quot;cell_name&quot;</span><span class="p">)</span>
<span class="n">colours</span><span class="o">=</span><span class="n">array</span><span class="p">([</span><span class="s2">&quot;r&quot;</span><span class="p">,</span><span class="s2">&quot;r&quot;</span><span class="p">,</span><span class="s2">&quot;r&quot;</span><span class="p">,</span><span class="s2">&quot;b&quot;</span><span class="p">,</span><span class="s2">&quot;b&quot;</span><span class="p">,</span><span class="s2">&quot;b&quot;</span><span class="p">])</span>
<span class="n">d</span><span class="o">=</span> <span class="n">cell_cycle_scores</span><span class="p">[[</span><span class="s1">&#39;G1.score&#39;</span><span class="p">,</span> <span class="s1">&#39;G2M.score&#39;</span><span class="p">,</span> <span class="s1">&#39;S.score&#39;</span><span class="p">,</span><span class="s1">&#39;S_score&#39;</span><span class="p">,</span> <span class="s1">&#39;G2M_score&#39;</span><span class="p">,</span><span class="s1">&#39;G1_score&#39;</span><span class="p">]]</span>
<span class="n">sns</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">corr</span><span class="p">(),</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;mako&quot;</span><span class="p">,</span>
               <span class="n">vmin</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">vmax</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">row_colors</span><span class="o">=</span><span class="n">colours</span><span class="p">,</span> <span class="n">col_colors</span><span class="o">=</span><span class="n">colours</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span> <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Pearson Correlation Heatmap&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">loc</span><span class="o">+</span><span class="s2">&quot;figures/heatmap.pdf&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/scripts_3_analysis_scripts_7_cellphase_model_11_0.png" src="../../_images/scripts_3_analysis_scripts_7_cellphase_model_11_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
<span class="n">cellcycle_dataframe</span><span class="o">=</span><span class="n">adata</span><span class="p">[:,</span><span class="n">gene_subset</span><span class="p">]</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">cellcycle_dataframe</span><span class="p">,</span> <span class="n">n_comps</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">cellcycle_dataframe</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">cellcycle_dataframe</span><span class="p">)</span>
<span class="n">cellcycle_dataframe</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">cellcycle_dataframe</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_umap&quot;</span><span class="p">],</span><span class="n">index</span><span class="o">=</span><span class="n">cellcycle_dataframe</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)</span>
<span class="n">Z_test</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">cellcycle_dataframe</span><span class="p">)</span>
<span class="n">Z_pred</span><span class="o">=</span><span class="n">knn</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">Z_test</span><span class="p">)</span>
<span class="c1">#subset the new dataframe</span>
<span class="n">cell_cycle_scores</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">knn</span><span class="o">.</span><span class="n">predict_proba</span><span class="p">(</span><span class="n">Z_test</span><span class="p">),</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;G1.score&quot;</span><span class="p">,</span><span class="s2">&quot;G2M.score&quot;</span><span class="p">,</span><span class="s2">&quot;S.score&quot;</span><span class="p">])</span>
<span class="c1">#set index to be S phase if score is &gt; 0.1</span>
<span class="n">cell_cycle_scores</span><span class="p">[</span><span class="s2">&quot;cell_barcode&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">Z_test</span><span class="o">.</span><span class="n">index</span>
<span class="n">cell_cycle_scores</span><span class="p">[</span><span class="s2">&quot;predicted_score&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">Z_pred</span>
<span class="n">cell_cycle_scores</span><span class="p">[</span><span class="s2">&quot;predicted_score&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/helenking/anaconda3/envs/apu/lib/python3.12/site-packages/scanpy/preprocessing/_pca.py:317: ImplicitModificationWarning: Setting element `.obsm[&#39;X_pca&#39;]` of view, initializing view as actual.
  adata.obsm[key_obsm] = X_pca
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
predicted_score
G1     11260
S       7455
G2M     4510
Name: count, dtype: int64
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#merge adata.obs and cell_cycle_scores on cellbarcode</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">cell_cycle_scores</span><span class="p">,</span> <span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s2">&quot;cell_barcode&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;cell_barcode&quot;</span><span class="p">)</span>
<span class="c1"># adata.obs=pd.concat([adata.obs.reset_index(drop=True), cell_cycle_scores], axis=1)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#change the shape of sns</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">rc</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;figure.figsize&#39;</span><span class="p">:(</span><span class="mi">5</span><span class="p">,</span><span class="mi">5</span><span class="p">)})</span>
<span class="n">d</span><span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="s1">&#39;S_score&#39;</span><span class="p">,</span> <span class="s1">&#39;G2M_score&#39;</span><span class="p">,</span> <span class="s1">&#39;G1.score&#39;</span><span class="p">,</span><span class="s1">&#39;G2M.score&#39;</span><span class="p">,</span> <span class="s1">&#39;S.score&#39;</span><span class="p">]]</span>
<span class="c1"># Compute the correlation matrix</span>
<span class="c1"># mask = np.triu(np.ones_like(d.corr(), dtype=bool))</span>
<span class="n">heatmap</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">corr</span><span class="p">(),</span> <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;BrBG&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/scripts_3_analysis_scripts_7_cellphase_model_14_0.png" src="../../_images/scripts_3_analysis_scripts_7_cellphase_model_14_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate G1 score as the ratio of G2M score to S score</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;G1_score&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;G2M_score&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;S_score&quot;</span><span class="p">]</span>

<span class="c1"># Define columns for phase assignment</span>
<span class="n">columns_phase_scanpy</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;guide_assignment&#39;</span><span class="p">,</span> <span class="s1">&#39;promoter_type&#39;</span><span class="p">,</span> <span class="s1">&#39;perturbation&#39;</span><span class="p">,</span> <span class="s1">&#39;guide_id&#39;</span><span class="p">,</span> <span class="s1">&#39;S_score&#39;</span><span class="p">,</span> <span class="s1">&#39;G2M_score&#39;</span><span class="p">,</span> <span class="s1">&#39;G1_score&#39;</span><span class="p">,</span> <span class="s1">&#39;phase&#39;</span><span class="p">]</span>

<span class="c1"># Create G1 score and define columns for phase assignment</span>
<span class="n">columns_phase_assignment</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;guide_assignment&#39;</span><span class="p">,</span> <span class="s1">&#39;successfulKD&#39;</span><span class="p">,</span> <span class="s1">&#39;promoter_type&#39;</span><span class="p">,</span> <span class="s1">&#39;perturbation&#39;</span><span class="p">,</span> <span class="s1">&#39;guide_id&#39;</span><span class="p">,</span> <span class="s1">&#39;S.score&#39;</span><span class="p">,</span> <span class="s1">&#39;G2M.score&#39;</span><span class="p">,</span> <span class="s1">&#39;G1.score&#39;</span><span class="p">,</span> <span class="s1">&#39;predicted_score&#39;</span><span class="p">]</span>
<span class="c1">#filter for :successfulKD</span>
<span class="n">phase_assignment</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">columns_phase_assignment</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
<span class="c1"># Save phase assignment to CSV</span>
<span class="n">phase_assignment</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">loc</span> <span class="o">+</span> <span class="s1">&#39;files/phase_assignment.csv&#39;</span><span class="p">)</span>

<span class="c1"># Define columns for calculating the number of cells in each phase and guide assignment</span>
<span class="n">columns_phase_scanpy_sub</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;guide_assignment&#39;</span><span class="p">,</span> <span class="s1">&#39;promoter_type&#39;</span><span class="p">,</span> <span class="s1">&#39;phase&#39;</span><span class="p">]</span>
<span class="n">columns_phase_assignment_sub</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;guide_assignment&#39;</span><span class="p">,</span> <span class="s1">&#39;promoter_type&#39;</span><span class="p">,</span> <span class="s1">&#39;predicted_score&#39;</span><span class="p">]</span>

<span class="c1"># Group by guide assignment and promoter type, then count the number of occurrences</span>
<span class="n">phase_assignment</span> <span class="o">=</span> <span class="n">phase_assignment</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">columns_phase_assignment_sub</span><span class="p">,</span> <span class="n">observed</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="n">phase_assignment</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;guide_assignment&#39;</span><span class="p">,</span> <span class="s1">&#39;promoter_type&#39;</span><span class="p">,</span> <span class="s1">&#39;phase&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]</span>

<span class="c1"># Extract target gene, guide number, and promoter type from guide assignment</span>
<span class="n">phase_assignment</span><span class="p">[</span><span class="s2">&quot;target_gene&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">phase_assignment</span><span class="p">[</span><span class="s2">&quot;guide_assignment&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">phase_assignment</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s2">&quot;guide_num&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">phase_assignment</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s2">&quot;guide_assignment&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<span class="n">phase_assignment</span><span class="p">[</span><span class="s2">&quot;promoter_type_2&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">phase_assignment</span><span class="p">[</span><span class="s2">&quot;guide_assignment&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

<span class="c1"># Subselect the non-targeting cells</span>
<span class="n">phase_assignment</span> <span class="o">=</span> <span class="n">phase_assignment</span><span class="p">[</span><span class="n">phase_assignment</span><span class="p">[</span><span class="s2">&quot;promoter_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">phase_assignment</span><span class="p">[</span><span class="s2">&quot;promoter_type_2&quot;</span><span class="p">]]</span>

<span class="c1"># Calculate the change in percentage of cells in each phase relative to non-targeting control</span>
<span class="c1"># Calculate the count percentage per guide assignment</span>
<span class="n">phase_assignment_phase</span> <span class="o">=</span> <span class="n">phase_assignment</span><span class="p">[[</span><span class="s1">&#39;guide_assignment&#39;</span><span class="p">,</span> <span class="s1">&#39;phase&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;guide_assignment&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span><span class="s1">&#39;count&#39;</span><span class="p">:</span> <span class="s1">&#39;sum&#39;</span><span class="p">})</span>
<span class="n">phase_assignment_phase</span> <span class="o">=</span> <span class="n">phase_assignment_phase</span><span class="p">[</span><span class="n">phase_assignment_phase</span><span class="p">[</span><span class="s2">&quot;count&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">]</span>

<span class="c1"># Merge the total count with the phase assignment data</span>
<span class="n">phase_assignment_phase</span> <span class="o">=</span> <span class="n">phase_assignment_phase</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">phase_assignment</span><span class="p">,</span> <span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s2">&quot;guide_assignment&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span> <span class="n">suffixes</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;_total&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">))</span>

<span class="c1"># Calculate the percentage of cells in each phase</span>
<span class="n">phase_assignment_phase</span><span class="p">[</span><span class="s2">&quot;percentage&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">phase_assignment_phase</span><span class="p">[</span><span class="s2">&quot;count&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="n">phase_assignment_phase</span><span class="p">[</span><span class="s2">&quot;count_total&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span>
<span class="n">phase_assignment_phase</span><span class="p">[</span><span class="s2">&quot;guide_num&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">phase_assignment_phase</span><span class="p">[</span><span class="s2">&quot;guide_num&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/var/folders/r_/bqnt2f_d6919cz7v1yzx2ykc0000gn/T/ipykernel_12504/1767973535.py:32: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  phase_assignment_phase = phase_assignment[[&#39;guide_assignment&#39;, &#39;phase&#39;, &#39;count&#39;]].groupby([&#39;guide_assignment&#39;]).agg({&#39;count&#39;: &#39;sum&#39;})
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">non_targeting</span><span class="o">=</span><span class="n">phase_assignment_phase</span><span class="p">[</span><span class="n">phase_assignment_phase</span><span class="p">[</span><span class="s2">&quot;promoter_type&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;Control&quot;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">non_targeting</span><span class="p">[</span><span class="s2">&quot;target_gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
<span class="n">ttest_list</span><span class="o">=</span><span class="p">[]</span>
<span class="k">for</span> <span class="n">phase</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;G2M&quot;</span><span class="p">,</span><span class="s2">&quot;S&quot;</span><span class="p">,</span><span class="s2">&quot;G1&quot;</span><span class="p">]:</span>
    <span class="n">non_targeting_phase</span><span class="o">=</span><span class="n">non_targeting</span><span class="p">[</span><span class="n">non_targeting</span><span class="p">[</span><span class="s2">&quot;phase&quot;</span><span class="p">]</span><span class="o">==</span><span class="n">phase</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">phase_assignment_phase</span><span class="p">[</span><span class="s2">&quot;target_gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">promoter</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;MP&quot;</span><span class="p">,</span><span class="s2">&quot;AP&quot;</span><span class="p">]:</span>
            <span class="c1">#ztest for each phase</span>
            <span class="c1">#subset the data for the gene and phase</span>
            <span class="n">gene_phase</span><span class="o">=</span><span class="n">phase_assignment_phase</span><span class="p">[(</span><span class="n">phase_assignment_phase</span><span class="p">[</span><span class="s2">&quot;target_gene&quot;</span><span class="p">]</span><span class="o">==</span><span class="n">gene</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">phase_assignment_phase</span><span class="p">[</span><span class="s2">&quot;promoter_type&quot;</span><span class="p">]</span><span class="o">==</span><span class="n">promoter</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">phase_assignment_phase</span><span class="p">[</span><span class="s2">&quot;phase&quot;</span><span class="p">]</span><span class="o">==</span><span class="n">phase</span><span class="p">)]</span>
            <span class="c1">#test them with</span>
            <span class="c1"># print(gene_phase.head())</span>
            <span class="k">if</span> <span class="n">gene_phase</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">==</span><span class="mi">2</span><span class="p">:</span>
                <span class="n">p1_perc</span><span class="o">=</span><span class="n">gene_phase</span><span class="p">[</span><span class="s2">&quot;percentage&quot;</span><span class="p">][</span><span class="n">gene_phase</span><span class="p">[</span><span class="s2">&quot;guide_num&quot;</span><span class="p">]</span><span class="o">==</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">p2_perc</span><span class="o">=</span><span class="n">gene_phase</span><span class="p">[</span><span class="s2">&quot;percentage&quot;</span><span class="p">][</span><span class="n">gene_phase</span><span class="p">[</span><span class="s2">&quot;guide_num&quot;</span><span class="p">]</span><span class="o">==</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">ss_score</span><span class="o">=</span><span class="n">num_sim</span><span class="p">(</span><span class="n">p1_perc</span><span class="p">,</span><span class="n">p2_perc</span><span class="p">)</span>
                <span class="n">ttest</span><span class="o">=</span><span class="n">ws</span><span class="o">.</span><span class="n">ttest_ind</span><span class="p">(</span><span class="n">gene_phase</span><span class="p">[</span><span class="s2">&quot;count&quot;</span><span class="p">],</span><span class="n">non_targeting_phase</span><span class="p">[</span><span class="s2">&quot;count&quot;</span><span class="p">],</span><span class="n">alternative</span><span class="o">=</span><span class="s2">&quot;two-sided&quot;</span><span class="p">,</span> <span class="n">usevar</span><span class="o">=</span><span class="s2">&quot;unequal&quot;</span><span class="p">)</span>
                <span class="n">ttest_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">gene</span><span class="p">,</span><span class="n">phase</span><span class="p">,</span><span class="n">promoter</span><span class="p">,</span><span class="n">ttest</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">ttest</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">ss_score</span><span class="p">])</span>

<span class="n">ttest_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">ttest_list</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;target_gene&quot;</span><span class="p">,</span><span class="s2">&quot;phase&quot;</span><span class="p">,</span><span class="s2">&quot;promoter_type&quot;</span><span class="p">,</span><span class="s2">&quot;pvalue&quot;</span><span class="p">,</span><span class="s2">&quot;tstat&quot;</span><span class="p">,</span><span class="s2">&quot;corr&quot;</span><span class="p">])</span>
<span class="n">ttest_df</span><span class="o">=</span><span class="n">ttest_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;pvalue&quot;</span><span class="p">)</span>
<span class="n">ttest_df</span><span class="p">[</span><span class="s2">&quot;significant&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">multi</span><span class="o">.</span><span class="n">fdrcorrection</span><span class="p">(</span><span class="n">ttest_df</span><span class="p">[</span><span class="s2">&quot;pvalue&quot;</span><span class="p">],</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;p&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">ttest_df</span><span class="p">[</span><span class="s2">&quot;adj_pvalue&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">multi</span><span class="o">.</span><span class="n">fdrcorrection</span><span class="p">(</span><span class="n">ttest_df</span><span class="p">[</span><span class="s2">&quot;pvalue&quot;</span><span class="p">],</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;p&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ttest_df</span><span class="p">[</span><span class="s2">&quot;significant&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">())</span>
<span class="n">non_targeting</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">loc</span><span class="o">+</span><span class="s2">&quot;files/non_targeting.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;non-targeting&#39;]
significant
False    230
True      31
Name: count, dtype: int64
</pre></div></div>
</div>
<p>Identifying Promoter-Specific Proliferation Defects The authors apply this trained model to their perturbed cell populations to see which alternative promoters control the “molecular switch” of cell division.</p>
<p>Statistical Enrichment: Using a two-sided z-test, the code identifies promoters that cause a significant shift in the distribution of cell phases compared to non-targeting controls.</p>
<p>Major Finding: The analysis reveals that 34.38% of alternative promoters significantly alter the cell cycle</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#merge ttest_df with phase_assignment_phase</span>
<span class="n">phase_assignment_phase</span><span class="o">=</span><span class="n">phase_assignment_phase</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">ttest_df</span><span class="p">)</span>
<span class="n">phase_assignment_phase</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">phase_assignment_phase</span><span class="p">[</span><span class="s2">&quot;target_gene&quot;</span><span class="p">]</span>
<span class="n">phase_assignment_phase</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">loc</span><span class="o">+</span><span class="s1">&#39;files/phase_assignment_phase.csv&#39;</span><span class="p">)</span>
<span class="n">phase_assignment_phase</span><span class="o">=</span><span class="n">phase_assignment_phase</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">adata_full</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span><span class="s2">&quot;successfulKD&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span> <span class="n">left_on</span><span class="o">=</span><span class="s2">&quot;target_gene&quot;</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>

<span class="n">successfulKJD</span><span class="o">=</span><span class="n">phase_assignment_phase</span><span class="p">[[</span><span class="s2">&quot;target_gene&quot;</span><span class="p">,</span><span class="s2">&quot;promoter_type&quot;</span><span class="p">,</span><span class="s2">&quot;significant&quot;</span><span class="p">]][</span><span class="n">phase_assignment_phase</span><span class="p">[</span><span class="s2">&quot;successfulKD&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;True&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
<span class="n">gene_sig</span><span class="o">=</span><span class="n">successfulKJD</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;target_gene&quot;</span><span class="p">,</span><span class="s2">&quot;promoter_type&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="c1">#print</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gene_sig</span><span class="p">[</span><span class="n">gene_sig</span><span class="p">[</span><span class="s2">&quot;significant&quot;</span><span class="p">]</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">/</span><span class="p">((</span><span class="n">gene_sig</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])))</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
0.34375
</pre></div></div>
</div>
<p>Divergent Effects: The model shows that knocking down the P1 promoter of ESR1 has minimal effect on cell growth, whereas knocking down the P2 promoter significantly arrests cells in a specific phase and reduces proliferation.</p>
<p>Mechanism: This result, visualized via heatmaps of “Spectra” activity scores, links alternative promoter usage directly to the physical behavior of breast cancer cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">successfulKJD</span><span class="o">=</span><span class="n">phase_assignment_phase</span><span class="p">[[</span><span class="s2">&quot;target_gene&quot;</span><span class="p">,</span><span class="s2">&quot;significant&quot;</span><span class="p">]][</span><span class="n">phase_assignment_phase</span><span class="p">[</span><span class="s2">&quot;successfulKD&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;True&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
<span class="n">gene_sig</span><span class="o">=</span><span class="n">successfulKJD</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;target_gene&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="c1">#print</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gene_sig</span><span class="p">[</span><span class="n">gene_sig</span><span class="p">[</span><span class="s2">&quot;significant&quot;</span><span class="p">]</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gene_sig</span><span class="p">[</span><span class="n">gene_sig</span><span class="p">[</span><span class="s2">&quot;significant&quot;</span><span class="p">]</span><span class="o">==</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(20, 1)
(12, 1)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#read adata back in and merge th adata.obs</span>
<span class="k">if</span> <span class="s2">&quot;predicted_score&quot;</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="n">adata_full</span><span class="o">=</span><span class="n">ad</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">loc</span><span class="o">+</span><span class="s2">&quot;files/adata_normalised.h5ad&quot;</span><span class="p">)</span>
    <span class="n">adata_full</span><span class="o">.</span><span class="n">obs</span><span class="o">=</span><span class="n">adata_full</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="s2">&quot;G1.score&quot;</span><span class="p">,</span><span class="s2">&quot;G2M.score&quot;</span><span class="p">,</span><span class="s2">&quot;S.score&quot;</span><span class="p">,</span><span class="s2">&quot;predicted_score&quot;</span><span class="p">]],</span> <span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
    <span class="c1">#write with cell cycle</span>
    <span class="c1">#change predicted_score from string to category</span>
    <span class="n">adata_full</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;predicted_score&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">adata_full</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;predicted_score&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;category&quot;</span><span class="p">)</span>
    <span class="n">adata_full</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="n">loc</span><span class="o">+</span><span class="s2">&quot;files/adata_normalised_cellcycle.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="6_perturbseq_tsne_kldivergence_umap.html" class="btn btn-neutral float-left" title="Quantitative individual UMAPs per gene of interest" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="8_perturbseq_differentialexpression.html" class="btn btn-neutral float-right" title="Differential Expression analysis between P1 &amp; P2" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2026, Helen King.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>